project/MimaExcludes.scala

```diff
@@ -241,7 +241,10 @@ object MimaExcludes {
 
     // [SPARK-26216][SQL] Do not use case class as public API (UserDefinedFunction)
     ProblemFilters.exclude[MissingClassProblem]("org.apache.spark.sql.expressions.UserDefinedFunction$"),
-    ProblemFilters.exclude[IncompatibleTemplateDefProblem]("org.apache.spark.sql.expressions.UserDefinedFunction")
+    ProblemFilters.exclude[IncompatibleTemplateDefProblem]("org.apache.spark.sql.expressions.UserDefinedFunction"),
+
+    // [SPARK-19591][ML] Add sample weights to decision trees
+    ProblemFilters.exclude[DirectMissingMethodProblem]("org.apache.spark.mllib.tree.configuration.Strategy.this")
```

Comment:
Although it's possible to break compatibility in Spark 3, this class appears to be a relatively stable API, for better or worse. It might be worth just retaining the old constructor. I don't know how much people use it though.

Reply:
How can I make this work while still having the new BeanProperty? I've tried creating a duplicate constructor but scala complained about duplicate signatures:
  /**
   * Backwards compatible constructor for [[org.apache.spark.mllib.tree.configuration.Strategy]]
   */
  @Since("1.0.0")
  def this(algo: Algo,
           impurity: Impurity,
           maxDepth: Int,
           numClasses: Int,
           maxBins: Int,
           quantileCalculationStrategy: QuantileStrategy = Sort,
           categoricalFeaturesInfo: Map[Int, Int] = Map[Int, Int](),
           minInstancesPerNode: Int = 1,
           minInfoGain: Double = 0.0,
           maxMemoryInMB: Int = 256,
           subsamplingRate: Double = 1,
           useNodeIdCache: Boolean = false,
           checkpointInterval: Int = 10
          ) {
    this(algo, impurity, maxDepth, numClasses, maxBins, quantileCalculationStrategy,
      categoricalFeaturesInfo, minInstancesPerNode, minInfoGain, maxMemoryInMB,
      subsamplingRate, useNodeIdCache, checkpointInterval, 0.0)
  }

Reply:
here is an example of the error:

[error]     val strategy = new Strategy(algo, impurity, maxDepth, numClasses, maxBins,
[error]                    ^
[error] /home/ilya/ilmat-spark/mllib/src/main/scala/org/apache/spark/mllib/tree/RandomForest.scala:170: overloaded method constructor Strategy with alternatives:
[error]   (algo: org.apache.spark.mllib.tree.configuration.Algo.Algo,impurity: org.apache.spark.mllib.tree.impurity.Impurity,maxDepth: Int,numClasses: Int,maxBins: Int,quantileCalculationStrategy: org.apache.spark.mllib.tree.configuration.QuantileStrategy.QuantileStrategy,categoricalFeaturesInfo: Map[Int,Int],minInstancesPerNode: Int,minInfoGain: Double,maxMemoryInMB: Int,subsamplingRate: Double,useNodeIdCache: Boolean,checkpointInterval: Int)org.apache.spark.mllib.tree.configuration.Strategy <and>
[error]   (algo: org.apache.spark.mllib.tree.configuration.Algo.Algo,impurity: org.apache.spark.mllib.tree.impurity.Impurity,maxDepth: Int,numClasses: Int,maxBins: Int,quantileCalculationStrategy: org.apache.spark.mllib.tree.configuration.QuantileStrategy.QuantileStrategy,categoricalFeaturesInfo: Map[Int,Int],minInstancesPerNode: Int,minInfoGain: Double,maxMemoryInMB: Int,subsamplingRate: Double,useNodeIdCache: Boolean,checkpointInterval: Int,minWeightFractionPerNode: Double)org.apache.spark.mllib.tree.configuration.Strategy
[error]  cannot be applied to (org.apache.spark.mllib.tree.configuration.Algo.Value, org.apache.spark.mllib.tree.impurity.Impurity, Int, Int, Int, org.apache.spark.mllib.tree.configuration.QuantileStrategy.Value, Map[Int,Int])

Reply:
Hm, yeah, the problem is that both constructors define a bunch of optional arguments, and this call doesn't specify those optional argument. So both constructors apply. The 'fix' would be to make the backwards-compatible constructor not take any optional arguments. It was already source-compatible without this additional constructor, but would make it binary-compatible too.

I'm on the fence about it. It's not something apps would normally specify directly, but it's also pretty trivial to write in this additional constructor. If it works, I'd keep the backwards-compatible constructor. Little benefit, very little cost.